lib/pc_sort.c

```diff
@@ -0,0 +1,267 @@
+/***********************************************************************
+* pc_sort.c
+*
+*  Pointclound patch sorting.
+*
+*  PgSQL Pointcloud is free and open source software provided
+*  by the Government of Canada
+*  Copyright (c) 2013 Natural Resources Canada
+*
+*  Author: M. Brédif
+*
+***********************************************************************/
+#include "pc_api_internal.h"
+#include <assert.h>
+#include <stdlib.h>
+
+
+/**
+ * Comparators
+ */
+
+int
+pc_compare_dim (const void *a, const void *b, void *arg)
+{
+    PCDIMENSION **dim = (PCDIMENSION **)arg;
+    uint32_t byteoffset     = dim[0]->byteoffset;
+    uint32_t interpretation = dim[0]->interpretation;
+    double da = pc_double_from_ptr(a+byteoffset,interpretation);
+    double db = pc_double_from_ptr(b+byteoffset,interpretation);
+    int cmp = ((da > db) - (da < db));
+    return ( cmp == 0 && dim[1]) ? pc_compare_dim(a,b,dim+1) : cmp;
```

Comment:
This assumption that `dim[1]` being zero implies that you're at the end of the array of dimensions... are you constructing this dimension list explicitly to include a null element at the end?